### PR TITLE
PID automatic tuning

### PR DESCRIPTION
**Pull request by greenblitz4590**

### CODE REVIEW

File: src/main/java/edu/greenblitz/robotName/utils/PIDObject.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        this(kp, ki, kd, 0.0, inv);
+    }
+
+    public PIDObject(double kp, double ki, int inv) {
+        this(kp, ki, 0.0, inv);
+    }
+
+    public PIDObject(double kp, int inv) {
+        this(kp, 0.0, inv);
+    }
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/PIDObject.java:71]: too much constructurs

File: src/main/java/edu/greenblitz/robotName/utils/PIDObject.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    public PIDObject withMaxPower(double maxPower) {
+        setMaxPower(maxPower);
+        return this;
+    }
+
+    public PIDController getPIDController() {
+        PIDController pidController = new PIDController(this.kp, this.ki, this.kd);
+        pidController.setTolerance(this.tolerance);
+        return pidController;
+    }
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/PIDObject.java:191]: line space end of class

File: src/main/java/frc/robot/RobotManager.java
Comment (YHerm) [inline src/main/java/frc/robot/RobotManager.java:1]: all changes here need to be deleted

File: src/main/java/frc/utils/calibration/autoPIDtuner/AutoPIDTuner.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    public PIDObject tunePIDController(double ku, double pu) {
+        return new PIDObject(ku / AutoPIDTunerConstants.PID_CONTROLLER.DIVIDE_KU_TO_GET_KP,
+                pu / AutoPIDTunerConstants.PID_CONTROLLER.DIVIDE_PU_TO_GET_KI,
+                pu / AutoPIDTunerConstants.PID_CONTROLLER.DIVIDE_PU_TO_GET_KD,
+                0, 0);
+    }
+
+}
+
+
```
Comment (YHerm) [inline src/main/java/frc/utils/calibration/autoPIDtuner/AutoPIDTuner.java:28]: remove these two last lines

File: src/main/java/frc/utils/calibration/autoPIDtuner/AutoPIDTuner.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package frc.utils.calibration.autoPIDtuner;
+
+import edu.greenblitz.robotName.utils.PIDObject;
+
+
+
```
Comment (YHerm) [inline src/main/java/frc/utils/calibration/autoPIDtuner/AutoPIDTuner.java:6]: these two spaces need to be gone

File: src/main/java/frc/utils/calibration/autoPIDtuner/AutoPIDTunerConstants.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package frc.utils.calibration.autoPIDtuner;
+
+public class AutoPIDTunerConstants {
```
Comment (YHerm) [inline src/main/java/frc/utils/calibration/autoPIDtuner/AutoPIDTunerConstants.java:3]: add space after this line

File: src/main/java/frc/utils/calibration/autoPIDtuner/ControllerType.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package frc.utils.calibration.autoPIDtuner;
+
+public enum ControllerType {
```
Comment (YHerm) [inline src/main/java/frc/utils/calibration/autoPIDtuner/ControllerType.java:3]: consider adding PD_CONROLLER

**Review by YHerm — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/utils/PIDObject.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+public class PIDObject {
+
+    private double kp, kd, ki, ff;
+
+    private double iZone;
+
+    private double tolerance;
+
+    private int inverted = 1;
```
Comment (hillelv) [inline src/main/java/edu/greenblitz/robotName/utils/PIDObject.java:13]: maybe change to boolean

File: src/main/java/frc/utils/calibration/autoPIDtuner/TunerPicker.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    private ControllerType controllerType;
+    AutoPIDTuner autoPIDTuner;
+
+    public TunerPicker(ControllerType controllerType) {
+        this.controllerType = controllerType;
+        this.autoPIDTuner = new AutoPIDTuner();
+    }
+
+    public PIDObject Tune(double ku) {
```
Comment (bossyshrimp68) [inline src/main/java/frc/utils/calibration/autoPIDtuner/TunerPicker.java:15]: functions start with lowercase, same below

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]
good job!
hillel and yoav made good comments, answer them too.

File: src/main/java/frc/utils/calibration/autoPIDtuner/ControllerType.java (hunk 6 above)
Comment (RJ0907) [reply src/main/java/frc/utils/calibration/autoPIDtuner/ControllerType.java:3]: that is the only type of controller that cannot be tuned automatically

File: src/main/java/edu/greenblitz/robotName/utils/PIDObject.java (hunk 7 above)
Comment (RJ0907) [reply src/main/java/edu/greenblitz/robotName/utils/PIDObject.java:13]: that is a code design. if we change it to boolean than we better rewrite this shitty code from the beginning

File: src/main/java/edu/greenblitz/robotName/utils/PIDObject.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+    private double maxPower;
+
+    public PIDObject() {
+        this(0);
+    }
+
+    public PIDObject(double kp, double ki, double kd, double kf, int inv) {
```
[the remark this answers is not in the document]
Comment (RJ0907) [reply src/main/java/edu/greenblitz/robotName/utils/PIDObject.java:21]: Read what I wrote to hillel

**Pull request closed without merging**